pac4j-saml/src/main/java/org/pac4j/saml/sso/impl/SAML2AuthnRequestBuilder.java

```diff
@@ -79,13 +82,15 @@ public SAML2AuthnRequestBuilder(final SAML2Configuration cfg) {
         this.providerName = cfg.getProviderName();
         this.extensions = cfg.getAuthnRequestExtensions();
         this.useNameQualifier = cfg.isUseNameQualifier();
+        this.urlResolver = urlResolver;
     }
 
     @Override
     public AuthnRequest build(final SAML2MessageContext context) {
         final SingleSignOnService ssoService = context.getIDPSingleSignOnService(this.bindingType);
         final String idx = this.assertionConsumerServiceIndex > 0 ? String.valueOf(assertionConsumerServiceIndex) : null;
         final AssertionConsumerService assertionConsumerService = context.getSPAssertionConsumerService(idx);
+        assertionConsumerService.setLocation(this.urlResolver.compute(assertionConsumerService.getLocation(), context.getWebContext()));
```

Comment:
This also doesn't look right to me. I am not a fan of overriding metadata behavior. Why not just change the ACS url based on what you need?

Reply:
Yes, I don't like it at all. This smells thread issues...

Let's be clear on this: if you want to handle multiple hosts/callback URLs, you'll need multiple SAML2 clients. Otherwise I don't think it will work properly (thread issues, one metadata...)

Reply:
Doh, I actually pushed the wrong commit - this was a preliminary attempt to check feasibility. This change must actually be performed on the `AuthnRequest` which is a local request-scoped transient object, not the `AssertionConsumerService`,  indeed!
What's the best way to handle this? Do you want me to create a brand new pull request or should I rebase this pull request to the next commit I'll push with the changes?

Reply:
Let's do that on this PR. You still have the issue of one metadata for N ACS URLs...

Reply:
I'm not sure I get this - but let's dig into it to make sure there is no hidden problem.

The code I committed since then performs the URL resolution only in the case when no `AssertionConsumerServiceIndex` is provided.
In other words, in case of an ACS metadata providing with multiple URLs *and* a `AssertionConsumerServiceIndex` is specified, then there is no URL location set nor resolved (even if configured).

In the other case, pac4j-saml relies on the configured callback URL - and this is where there is now URL resolution taking place.

Is there any missing case which I'm blind to?
Could you have a look at the latest committed code and let me know what you think?

Reply:
Manipulating ACS urls dynamically is a violation of the specification and the deployment profile:

> The AssertionConsumerServiceURL value, if present, MUST match an endpoint location expressed in the SP’s metadata exactly, without requiring URL canonicalization/normalization.

Reply:
We should have started here in fact ;-)

Among all things, we won't go against the the spec for sure. Even if this trick was appealing.